### PR TITLE
Input argument validation layer

### PR DESCRIPTION
**Pull request by MathiasMagnus**
# Input Parameter Validation Layer

_(This PR is based on top of #7 that's why the changeset is currently large, but folders cleanly separate changes.)_

The layer detects and reports (almost) all errors on input parameters and log, for each OpenCL API entry point of base OpenCL 3.0 specification (without extensions), (almost) all error conditions as defined in the API documentation that derive from the parameters used to call the function, and optionally returns the error. All the parameter tests are described in XML with the help of schema below.

The layer is stateless as of now and rely on querying properties of OpenCL objects from the underlying OpenCL implementation (or layer). The sequence of parameter validation tests is structured in such a way that querying should proceed only on the existing objects with guaranteed type and properties.

> Note: the stateless nature is a simplification only. Keeping track of object relationships is very much possible and the bulk of it has already been done in the object lifetime layer. Reusing that logic can be added later, but it does requre mild refactoring. Doing so would still require some queries be done on the fly or during init, because some validations depend on runtime properties, such as device supporting images or not, as these alter the validity of input args to some functions.

## Source files

The layer depends on an XML library of our choice ([RapidXML](http://rapidxml.sourceforge.net/)) which is used to compile `param-verification/parser.cpp`, which results in a binary that processes `param-verification/cl-avl.xml`, a copy of upstream `cl.xml` from OpenCL-Docs. This binary generates the rest of the source file `res.cpp` which together with the auxiliary:
- `param_verification.hpp`
- `object_is_valid.cpp`
- `list_violation.cpp`
- `struct_violation.cpp`
helper files comprise the bulk of the validation logic. The initialization of the layer is in the file `param-verification/param_verification.cpp`.

CMake build automation properly takes care of both fetching RapidXML (if not found) and also orders the building of the code generator with the rest of the build.

## XML schema
Starting from [cl.xml](https://github.com/KhronosGroup/OpenCL-Docs/blob/main/xml/cl.xml) of OpenCL, we use the definitions of functions of the type

```xml
        <command suffix="CL_API_SUFFIX__VERSION_1_0">
            <proto><type>cl_mem</type>                                  <name>clCreateBuffer</name></proto>
            <param><type>cl_context</type>                              <name>context</name></param>
            <param><type>cl_mem_flags</type>                            <name>flags</name></param>
            <param><type>size_t</type>                                  <name>size</name></param>
            <param><type>void</type>*                                   <name>host_ptr</name></param>
            <param><type>cl_int</type>*                                 <name>errcode_ret</name></param>
```

and extend them with conditions for violations of argument validity with corresponding returned errors. Conditions follow the the parameters and are enclosed by `<conditions>` tag. Each condition consists of two tags: `<if>` and `<then>`. The former represents arguments invalidation condition itself, the latter provides outcomes, i.e. specific function returns.

`<if>` tag consists of conditions on parameters that invalidate function in some way. Tags of `<not>`, `<and>`, `<or>` can be used with usual meaning to combine simpler conditions, `<and>` and `<or>` can include two or more conditions inside. Basic conditions are:
* `<eq>` - equal
* `<neq>` - not equal
* `<gt>` - greater
* `<ls>` - less
* `<array_len_ls>` - length of array provided by pointer xxx is less then yyy (dummy function as of yet, always provides false)
* `<bit_and>` - gives true if bitwise and operation between xxx and yyy returns non-zero
* `<mutex_violation>` - gives true if two or more of conditions inside are true
* `<from>` - gives true if version of OpenCL is greater or equal to xxx
* `<enum_violation name="xxx"/>` - gives true if the value of parameter xxx is outside of valid for its type 
* `<bitfield_violation name="xxx"/>` - gives true if the value of parameter xxx is outside of valid for its type (0 is treated as valid)
* `<list_violation name="xxx" param="yyy, zzz"/>` - gives true if the structure of list xxx is invalid for its type taking into account also the values of yyy, zzz, etc. (param property is optional, test functions should be written manually for the layer)
* `<struct_violation name="xxx" param="yyy, zzz"/>` - gives true if the structure of xxx is invalid taking into account also the values of yyy, zzz, etc. (param property is optional, test functions should be written manually)
* `<any_nullptr array="xxx" elements="yyy"/>` - gives true if any of the pointers in array xxx of yyy elements is NULL
* `<any_zero array="xxx" elements="yyy"/>` - gives true if any of the values in array xxx of yyy elements is 0
* `<object_is_invalid name="xxx"/>` - gives true if the object is invalid
* `<object_is_invalid name="xxx" type="yyy"/>` - gives true if the object is invalid or is not of the given type
* `<any_invalid array="xxx" elements="yyy"/>` - gives true if any of the objects in array xxx of yyy elements is invalid
* `<any_non_null_invalid array="xxx" elements="yyy"/>` - gives true if any of the objects in array xxx of yyy elements besides its NULL elements is invalid
* `<any_not_available array="devices" elements="yyy"/>` - gives true if any of yyy devices is unavailable
* `<object_not_in object="xxx" in="yyy"/>` - gives true if object xxx does not belong to object yyy (e.g. queue is not on device, contexts of the xxx and yyy are not the same, etc.)
* `<any_object_not_in array="xxx" elements="yyy" in="zzz"/>` - gives true if any object in array xxx of yyy elements does not belong to object zzz (e.g. contexts of the events xxx and command queue zzz are not the same, etc.)
* `<for_all in="xxx" query="yyy">` - gives true if all the evaluations of expression inside the tag for all devices associated with xxx (e.g. all devices in context, associated with kernel, etc.) give true, expression can use `<name>query</name>` as value of queried property yyy for every device 
* `<for_any in="context" query="xxx">` - gives true if any of the evaluations of expression inside the tag for all devices associated with xxx (e.g. all devices in context, associated with kernel, etc.) give true, expression can use `<name>query</name>` as value of queried property yyy for every device 
* `<not_aligned pointer="xxx" align="yyy"/>` - gives true if pointer xxx is not aligned to yyy bytes
* `<check_copy_overlap param="xxx"/>` - gives true is provided parameters xxx ensure overlap of regions to be read from and written to

Parameters of the conditions can be of types: `<name>`, `<literal>`, `<literal_list>`, `<sizeof>`, `<mult>`, `<mod>`, `<max>`, `<query>`. 
* `<name>` gives the value of the corresponding function parameter.
* `<literal>` gives literal constant.
* `<literal_list>` gives literal constant from the list connected to the type of parameter, corresponding to the value of parameter inside the tag.
* `<sizeof>` gives size of type.
* `<mult>` provides multiplication of its parameters.
* `<add>` provides addition of its parameters.
* `<mod>` gives xxx modulo yyy.
* `<max>` gives maximum of its two parameters.
* `<query object="xxx" property="yyy"/>` gives the value of property yyy of object xxx (or of the associated object, i.e. you can query property of device associated with command queue xxx), if it is a variable-size array property, only the first array element is returned

Example (continued):
```xml
<if>
  <for_all in="context" query="CL_DEVICE_MAX_MEM_ALLOC_SIZE">
    <gt>
      <name>size</name> <name>query</name>
    </gt>
  </for_all>
</if>
```

`<then>` tag consists of blocks defining returned parameters for the specific case defined in the preceding `<if>` tag and non-mandatory `<log>` tag that contains information on the parameter violation. Each block consists of two sequential tags: `<name>` and `<value>`. The former provides parameter name (including the function itself), the latter provides value returned.

Example (continued):
```xml
  <then>
    <log>size is greater than CL_DEVICE_MAX_MEM_ALLOC_SIZE for all devices in context</log>
    <name>clCreateBuffer</name> <value>NULL</value>
    <name>errcode_ret</name>    <value>CL_INVALID_BUFFER_SIZE</value>
  </then>
</command>
```

## Layer tests

Layer tests are set up in a similar way to the object lifetime tests. In fact, the common testing logic was extracted into separate files so that it does not need to be written twice.

### Implementation notes

The source files for the common code for testing can be found in utils/test/. The setup is as follows:
- `utils/test/CMakeLists.txt` defines the interface LayersTest. Tests can link with this to get the common functionality.
- `utils/test/layers_test.hpp` contains the common code, in namespace `layers_test`. This consists of the `setup` and `finalize` calls to set up and tear down the test, including parsing opencl versions and such, and a simple but extensible argument parsing system. This file also defines a basic version of the `EXPECT_SUCCESS` and `EXPECT_ERROR` macros. It defines the following interface, to be specialized for any particular test:
  - `layers_test::TestContext<TestOptions>` provides the basic context that keeps track of test state and common options. `TestOptions` provides test-specific options, this type should have the member function `bool parseArg(int argc, char* argv[], int& i);`. This function should attempt to parse argument `argv[i]`, and increments it count *with the number of extra parameters* (so arguments that dont have values don't need to increment this count).
  - `layers_test::TestContext<TestOptions>::fail()` registers that a check has failed. The test is considered failing if the total fail count is > 0.
  - `layers_test::TestContext<TestOptions>::finalize()` is used to finalize the test. Returns an exit code.
  - `layers_test::TestContext<TestOptions>::version` is the OpenCL version that the test is running under
  - `layers_test::TestContext<TestOptions>::platform` is the OpenCL platform name string to test with. It is passed for all tests using the option `--platform <name>`. The name needs to be an exact match, and defaults to the name of the object-lifetime test ICD.
  - `layers_test::TestContext<TestOptions>::options` allows access to the `TestOptions` for a particular test.
  - `layers_test::setup<TestOptions>(int argc, char* argv[], cl_version required_version, cl_platform_id& platform, cl_device_id& device)` is used to setup a test. It returns an initialized test context, and the selected platform and device to use for testing.
  - `std::ostream& layers_test::log(const char* file, int line)` returns a log stream intended for printing some information related to a particular line. The stream is in fact just cout.
  - `LAYERS_TEST_LOG()` does the same but has the file and line filled in to the current ones.
  - `LAYERS_TEST_EXPECT_SUCCESS(context, status)` checks that `status` is `CL_SUCCESS`, and registeres a failure otherwise.
  - `LAYERS_TEST_EXPECT_ERROR(context, status, error)` checks that `status` is `error`, and registeres a failure otherwise.

The actual tests for the parameter verification layer are found in `param-verification/test/`. The common testing system is transformed into a specialized version with a similar interface as the previous object lifetime tests by param-verification/test/param_verification_test.hpp
- `param_verification_test::setup()` is used to initialize the test as previous.
- `param_verification_test::finalize()` is used to finalize the test, and returns an exit code.
- `param_verification_test::TEST_CONTEXT` is the specialized test context to use for this test. Its `TestOptions` define no additional options.
- `EXPECT_SUCCESS(status)` is the same as `LAYERS_TEST_EXPECT_SUCCESS` except that it passes `param_verification_test::TEST_CONTEXT` for you.
- `EXPECT_ERROR(status, error)` does the same.

Additionally, the object lifetime tests have been modified to use the same system. This entails the following changes:
- The namespace has changed from `layer_test` to `object_lifetime_test`.
- Options like `ref_count_includes_implicit` are now accessed through `object_lifetime_test::TEST_CONTEXT.options`

The parameter verification tests also use the object-lifetime test ICD. To this end, it has been modified to accept the environment variable `ALWAYS_RETURN_SUCCESS`, which, as the name suggests, makes the ICD return `CL_SUCCESS` unconditionally. This enables us to test by feeding invalid parameters to OpenCL, and then checking whether the parameter verification layer correctly intercepted the call and made it return the appropriate error code.

> Note: we couldn't tell if the minor changes made to CLObjectLifetimeICD warranted extracting it as a common utility. There is a tradeoff between keeping mock ICDs as simple as reasonably possible and deduplicating/specializing logic toward the tested layer at hand. As follow-up features the mock ICD could be extracted to be shared between all layers, or even be considered to be shared with the ICD Loader mock as well.

In general, the tests are structured in a similar way, with each source file in `param-verification/test/*.cpp` having a corresponding `param-verification/test/*.regex` that contains the expected log output.

> Note: one notable exception in the testing system is `param-verification/test/object_validity.cpp`. Currently, the layer uses the backing implementation or layers to tell whether an object is valid, but due to a complexity in the ICD loader this does not quite work without the object-lifetime layer to prevent calls with invalid values being dispatched to the implementation. This is because the ICD loader uses a separate dispatch table for the first layer, rather than the object's dispatch table. The object's dispatch table is still filled with the function pointers of the implementation, which means that when `tdispatch->clSomeFunction(handle, ...)` is invoked, the ICD loader will attempt to perform `handle->dispatch->clSomeFunction` if this is the last layer in the chain. (This required noisy CI updates on Windows.)

### CODE REVIEW

**Comment by Kerilk** [comment]
@MathiasMagnus kind reminder to rebase this one so comparison is a tad easier.

File: object-lifetime/object_lifetime_test_icd.hpp [hunk 1]
```diff
@@ -12,13 +12,13 @@
 #include <map>          // std::map<std::string, void*> _extensions
 #include <set>          // std::set<std::shared_ptr<_cl_object>> _objects
 #include <utility>      // std::make_pair
-#include <mutex>        // std::mutex
```
Comment (Kerilk) [inline object-lifetime/object_lifetime_test_icd.hpp:15]: Is this removal intended?

File: object-lifetime/object_lifetime_test_icd.cpp [hunk 2]
```diff
@@ -1448,7 +1864,6 @@ void _cl_platform_id::init_dispatch()
   dispatch->clRetainProgram = clRetainProgram_wrap;
   dispatch->clReleaseProgram = clReleaseProgram_wrap;
   dispatch->clCreateKernel = clCreateKernel_wrap;
-  dispatch->clCreateKernelsInProgram = clCreateKernelsInProgram_wrap;
```
Comment (Kerilk) [inline object-lifetime/object_lifetime_test_icd.cpp:1451]: This seems to remove the most recent additions from the lifetime counting layer.

File: object-lifetime/object_lifetime_test_icd.cpp [hunk 3]
```diff
@@ -1361,7 +1770,6 @@ _cl_platform_id::_cl_platform_id()
   , _kernels{}
   , _events{}
   , _samplers{}
-  , _global_mutex{}
```
Comment (Kerilk) [inline object-lifetime/object_lifetime_test_icd.cpp:1364]: Is this removal intended?

File: object-lifetime/object_lifetime_test_icd.cpp [hunk 4]
```diff
@@ -1029,52 +1484,6 @@ cl_kernel _cl_program::clCreateKernel(
...
-      result.cend(),
-      kernels,
-      [](const std::shared_ptr<_cl_kernel>& kernel){ return kernel.get(); }
-    );
-    reference(kernel_count);
-  }
-
-  return CL_SUCCESS;
-}
-
```
Comment (Kerilk) [inline object-lifetime/object_lifetime_test_icd.cpp:1077]: This seems to remove the most recent additions from the lifetime counting layer.

File: object-lifetime/object_lifetime.cpp [hunk 5]
```diff
@@ -2564,7 +2554,7 @@ static CL_API_ENTRY cl_int CL_API_CALL clGetDeviceIDsFromD3D10KHR_wrap(
     devices,
     num_devices);
   cl_uint actual_num_entries = std::min(*num_devices, num_entries);
-  if (devices && result == CL_SUCCESS && actual_num_entries > 0)
+ if (devices && result == CL_SUCCESS && actual_num_entries > 0)
```
Comment (Kerilk) [inline object-lifetime/object_lifetime.cpp:2557]: This seems to remove the most recent additions from the lifetime counting layer.

File: cmake/run_and_compare.cmake [hunk 6]
```diff
@@ -18,13 +24,10 @@ if(EXPECTED_OUTPUT)
...
 if(EXTRA_OUTPUT)
-    file(READ "${EXTRA_OUTPUT}"          OUTPUT_FILE)
     file(READ "${EXPECTED_EXTRA_OUTPUT}" EXPECTED_FILE)
 
-    if(EXPECTED_FILE) # More friendly towards empty regexes
-        if(NOT "${OUTPUT_FILE}" MATCHES "^${EXPECTED_FILE}$")
-            message(FATAL_ERROR "Mismatch in output: ${OUTPUT_FILE}, expected: ${EXPECTED_FILE}")
-        endif()
+    if(NOT "${OUTPUT_FILE}" MATCHES "^${EXPECTED_FILE}$")
+        message(FATAL_ERROR "Mismatch in output: ${OUTPUT_FILE}, expected: ${EXPECTED_FILE}")
```
Comment (Kerilk) [inline cmake/run_and_compare.cmake:30]: This seems to remove the most recent additions from the lifetime counting layer.

**Review by Kerilk — CHANGES_REQUESTED** [review]

**Comment by Kerilk** [comment]
This first round of comments addresses the rebase which seems problematic:
https://github.com/KhronosGroup/OpenCL-Layers/pull/7/commits/5958fec53573da41f43897c5cf27bd9b44401350 https://github.com/KhronosGroup/OpenCL-Layers/pull/7/commits/78f40fceb59f8810cef3d47cd7b93b1978e5bede https://github.com/KhronosGroup/OpenCL-Layers/pull/7/commits/8bc5c9b81a45016d04e486d430508ae0f18e4991 https://github.com/KhronosGroup/OpenCL-Layers/pull/7/commits/cc32130f90bf985f43898b4fe1034b1c5fa16f5b
seem to have been partially or totally lost after the rebase. I tried to pinpoint all the problematic sections I could find, but some changes may have been intentional, so a thorough verification is required. Sorry about that.

**Comment by MathiasMagnus** [comment]
@Kerilk Yeah, apologies for that. I added those missing commits on top plus one correcting the namespaces of the test helpers.

File: cmake/run_and_compare.cmake (hunk 6 above)
Comment (MathiasMagnus) [reply cmake/run_and_compare.cmake]: In the now properly rebased version this seems to be solved.

**Comment by Kerilk** [comment]
One small comment:
 * For Infos that return a variable amount of data, shouldn't we be querying the underlying driver for the size of the data?
Example with `clGetDeviceInfo` and `CL_DEVICE_NAME` for now the layer is comparing the size of the provided buffer with `sizeof(char[1])`. To be valid, shouldn't the buffer be of at least the size the underlying implementation requested, obtained through the `param_value_size_ret`? This problem is different than the one mentioned in https://github.com/KhronosGroup/OpenCL-Layers/issues/12 and does not require as much tooling.

**Comment by Kerilk** [comment]
Aside from the small comment above, things look good from here.

**Review by Kerilk — APPROVED** [review]

**Pull request merged**